src/util/merge_irep.cpp

```diff
@@ -145,18 +145,26 @@ void merge_irept::operator()(irept &irep)
 
 const irept &merge_irept::merged(const irept &irep)
 {
+  if(irep.get_sub().empty() && irep.get_named_sub().empty())
```

Comment:
How about we start with an unconditional `insert` regardless of `get_sub()` or `get_named_sub()`, then alter it in place if any of the child merges do anything?

Reply:
Let's assume `irept` _A_ was previously inserted into the store. Now we have an `irept` _B_. If the unconditional insert of _B_ is unsuccessful (because _A_==_B_), then each of the child merges would result in a `detach`, meaning that sharing between _A_ and its stored copy is now broken. At the end, _B_ would now be the shared copy of the store, and so it would continue every time you call this method. We would actually not have any sharing at all anymore.

Reply:
Huh? If the insert is unsuccessful then only read-only operations are performed (`operator==` and `hash`), after which a reference to *A* is returned. You then return that, like the `find` call currently immediately below the new `insert`

Reply:
Ah, sorry, I didn't get that you still wanted to return immediately. Let me try to see how that would perform.

Reply:
Constness of the inserted element gets in the way I'm afraid: the store is a `std::unordered_set`, so elements once inserted cannot be modified in place.

Reply:
Personally I'd say this is an ideal case for use of `const_cast`, since we're making changes that don't affect its `==`-ness or its hash code. Also it's surely worth it to avoid the `find` then `insert` that we get for ireps not currently in the store?

Reply:
I've got an implementation that does what you proposed, but this now highlights how broken our use of `parametert` is: sometimes two `parametert` with different `ID_C_identifier` should be considered distinct (when using them to assign to arguments), while otherwise we would want them to be the same (while type checking declarations vs definitions). I'll see what I can do...

Reply:
Although the `parametert` business should still be cleaned up, it was actually me being stupid. I should have benchmarking results in a few hours.